docs/guides/deno/byo-deep-research.mdx

```diff
@@ -0,0 +1,738 @@
+// Configuration from environment variables
+const LLM_CONFIG = {
+  baseURL: Deno.env.get('LLM_BASE_URL') || 'http://localhost:11434/v1',
+  apiKey: Deno.env.get('LLM_API_KEY') || 'ollama',
+  model: Deno.env.get('LLM_MODEL') || 'llama3.2:3b',
+}
+
+const OAI = new OpenAI({
+  baseURL: LLM_CONFIG.baseURL,
+  apiKey: LLM_CONFIG.apiKey,
+})
+
+// Example .env file:
+// LLM_BASE_URL=http://localhost:11434/v1   # For local testing with Ollama
+// LLM_BASE_URL=https://api.openai.com/v1   # For production with OpenAI
+// LLM_API_KEY=your-api-key                 # For local this can be any value
+// LLM_MODEL=gpt-4-turbo-preview            # For production (or any other model of your choosing)
+// LLM_MODEL=llama3.2:3b                    # For local testing
+```
+
+### Prompt Templates
+
+The system uses three types of LLM calls, each with its own prompt template:
+
+1. **Query Generation** (`queryPrompt`): Generates search queries based on the research topic
+
+```typescript title:prompts/query.ts
+export default (date: string, topic: string) => `
+Your goal is to generate a targeted web search query.
+
+<CONTEXT>
+Current date: ${date}
+Please ensure your queries account for the most current information available as of this date.
+</CONTEXT>
+
+<TOPIC>
+${topic}
+</TOPIC>
+
+<FORMAT>
+Format your response as a JSON object with ALL two of these exact keys:
+   - "query": The actual search query string
+   - "rationale": Brief explanation of why this query is relevant
+</FORMAT>
+
+<EXAMPLE>
+Example output:
+{
+    "query": "machine learning transformer architecture explained",
+    "rationale": "Understanding the fundamental structure of transformer models"
+}
+</EXAMPLE>
+
+Provide your response only in JSON format:
+`
+```
+
+2. **Content Summarization** (`summarizerPrompt`): Summarizes content in the context of research topics
+
+```typescript title:prompts/summarizer.ts
+export default (topics: string[]) => `<GOAL>
+Generate a high-quality summary of the provided context for the following topics: 
+${topics.map((topic) => `- ${topic}`).join('\n')}
+</GOAL>
+
+<REQUIREMENTS>
+When creating a NEW summary:
+1. Highlight the most relevant information related to the user topic from the search results
+2. Ensure a coherent flow of information
+
+When EXTENDING an existing summary:
+1. Read the existing summary and new search results carefully.
+2. Compare the new information with the existing summary.
+3. For each piece of new information:
+    a. If it's related to existing points, integrate it into the relevant paragraph.
+    b. If it's entirely new but relevant, add a new paragraph with a smooth transition.
+    c. If it's not relevant to the user topic, skip it.
+4. Ensure all additions are relevant to the user's topic.
+5. Verify that your final output differs from the input summary.
+</REQUIREMENTS>
+
+<FORMATTING>
+- Start directly with the updated summary, without preamble or titles. Do not use XML tags in the output.
+</FORMATTING>
+
+<Task>
```

Comment:
Not sure on the syntax for this context definition, but can <Task> be capitalised for consistency, i.e. <TASK>?

Reply:
Done